File: Inventor-Training-Material-0c232f96f36e1e0bed31b07a6a6ebb901a31dd40/Module 09 - User Interface/RibbonDemo - VB.Net/VBNetRibbonDemo/StandardAddInServer.vb
Imports Inventor

Imports System.Drawing
Imports System.Runtime.InteropServices
Imports Microsoft.Win32
Imports System.Windows.Forms
Imports stdole


Namespace VBNetRibbonDemo
    <ProgIdAttribute("VBNetRibbonDemo.StandardAddInServer"), _
    GuidAttribute("af480230-9c8c-435e-a827-6f520e11b97b")> _
    Public Class StandardAddInServer
        Implements Inventor.ApplicationAddInServer

        ' Inventor application object.
        Private m_inventorApplication As Inventor.Application

        Private mAsmButtonDef As ButtonDefinition
        Private mPartButtonDef As ButtonDefinition

        Private Const strAddInGuid As String = "af480230-9c8c-435e-a827-6f520e11b97b"

#Region "ApplicationAddInServer Members"

        Public Sub Activate(ByVal addInSiteObject As Inventor.ApplicationAddInSite, ByVal firstTime As Boolean) Implements Inventor.ApplicationAddInServer.Activate

            ' This method is called by Inventor when it loads the AddIn.
            ' The AddInSiteObject provides access to the Inventor Application object.
            ' The FirstTime flag indicates if the AddIn is loaded for the first time.

            ' Initialize AddIn members.
            m_inventorApplication = addInSiteObject.Application

            ' TODO:  Add ApplicationAddInServer.Activate implementation.
            ' e.g. event initialization, command creation etc.

            Dim oCtrlDefs As ControlDefinitions = m_inventorApplication.CommandManager.ControlDefinitions

            Dim oStream1 As System.IO.Stream = Me.GetType().Assembly.GetManifestResourceStream("VBNetRibbonDemo.Tools.ico")
            Dim oIcon1 As Icon = New Icon(oStream1)
            Dim oIPictureDisp1 As Object = AxHostConverter.ImageToPictureDisp(oIcon1.ToBitmap())

            mAsmButtonDef = oCtrlDefs.AddButtonDefinition("Assembly Cmd", _
                                                          "Autodesk:RibbonVBTest:Button1", _
                                                          CommandTypesEnum.kQueryOnlyCmdType, _
                                                          strAddInGuid, _
                                                          "Description", _
                                                          "Tolltip text", _
                                                          oIPictureDisp1, _
                                                          Nothing, _
                                                          ButtonDisplayEnum.kDisplayTextInLearningMode)

            mPartButtonDef = oCtrlDefs.AddButtonDefinition("Part Cmd", _
                                                          "Autodesk:RibbonVBTest:Button2", _
                                                          CommandTypesEnum.kQueryOnlyCmdType, _
                                                          strAddInGuid, _
                                                          "Description", _
                                                          "Tolltip text", _
                                                          oIPictureDisp1, _
                                                          Nothing, _
                                                          ButtonDisplayEnum.kDisplayTextInLearningMode)


            If firstTime Then

                Dim UIManager As UserInterfaceManager = m_inventorApplication.UserInterfaceManager

                If m_inventorApplication.UserInterfaceManager.InterfaceStyle = InterfaceStyleEnum.kRibbonInterface Then

                    ' Get the assembly ribbon.
                    Dim assemblyRibbon As Inventor.Ribbon = UIManager.Ribbons.Item("Assembly")

                    ' Get the "Assemble" tab.
                    Dim assemblyTab As Inventor.RibbonTab = assemblyRibbon.RibbonTabs.Item("id_TabAssemble")

                    ' Create a new panel on the Assemble tab.
                    Dim panel1 As Inventor.RibbonPanel = assemblyTab.RibbonPanels.Add("Custom commands", _
                                                                                     "Autodes:RibbonVBTest:Panel1", _
                                                                                     strAddInGuid)
                    panel1.CommandControls.AddButton(mAsmButtonDef, True)



                    ' Get the part ribbon.
                    Dim partRibbon As Inventor.Ribbon = UIManager.Ribbons.Item("Part")

                    ' Get the "Model" tab.
                    Dim modelTab As Inventor.RibbonTab = partRibbon.RibbonTabs.Item("id_TabModel")

                    ' Create a new panel on the model tab.
                    Dim panel2 As Inventor.RibbonPanel = modelTab.RibbonPanels.Add("Custom commands", _
                                                                                     "Autodes:RibbonVBTest:Panel2", _
                                                                                     strAddInGuid)
                    panel2.CommandControls.AddButton(mPartButtonDef, True)

                Else

                    Dim oAsmCommandBar As CommandBar = UIManager.CommandBars("AMxAssemblyPanelCmdBar")
                    oAsmCommandBar.Controls.AddButton(mAsmButtonDef, 0)

                    Dim oPartCommandBar As CommandBar = UIManager.CommandBars("PMxPartFeatureCmdBar")
                    oPartCommandBar.Controls.AddButton(mPartButtonDef, 0)

                End If

            End If

            AddHandler mAsmButtonDef.OnExecute, AddressOf Me.mAsmButtonDef_OnExecute
            AddHandler mPartButtonDef.OnExecute, AddressOf Me.mPartButtonDef_OnExecute

        End Sub

        '//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        '// Use: Assembly Button Handler
        '//
        '//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        Private Sub mAsmButtonDef_OnExecute(ByVal Context As Inventor.NameValueMap)

            System.Windows.Forms.MessageBox.Show("Assembly button was clicked!!", "Assembly Button")

        End Sub

        '//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        '// Use: Part Button Handler
        '//
        '//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
        Private Sub mPartButtonDef_OnExecute(ByVal Context As Inventor.NameValueMap)

            System.Windows.Forms.MessageBox.Show("Part button was clicked!!", "Part Button")

        End Sub


        Public Sub Deactivate() Implements Inventor.ApplicationAddInServer.Deactivate

            ' This method is called by Inventor when the AddIn is unloaded.
            ' The AddIn will be unloaded either manually by the user or
            ' when the Inventor session is terminated.

            ' TODO:  Add ApplicationAddInServer.Deactivate implementation

            ' Release objects.
            Marshal.ReleaseComObject(m_inventorApplication)
            m_inventorApplication = Nothing

            System.GC.WaitForPendingFinalizers()
            System.GC.Collect()

        End Sub

        Public ReadOnly Property Automation() As Object Implements Inventor.ApplicationAddInServer.Automation

            ' This property is provided to allow the AddIn to expose an API 
            ' of its own to other programs. Typically, this  would be done by
            ' implementing the AddIn's API interface in a class and returning 
            ' that class object through this property.

            Get
                Return Nothing
            End Get

        End Property

        Public Sub ExecuteCommand(ByVal commandID As Integer) Implements Inventor.ApplicationAddInServer.ExecuteCommand

            ' Note:this method is now obsolete, you should use the 
            ' ControlDefinition functionality for implementing commands.

        End Sub

    

#End Region

        ''from http://blogs.msdn.com/b/andreww/archive/2007/07/30/converting-between-ipicturedisp-and-system-drawing-image.aspx

        Friend Class AxHostConverter
            Inherits AxHost
            Private Sub New()
                MyBase.New("")
            End Sub


            Public Shared Function ImageToPictureDisp(image As Image) As stdole.IPictureDisp
                Return DirectCast(GetIPictureDispFromPicture(image), stdole.IPictureDisp)
            End Function


            Public Shared Function PictureDispToImage(pictureDisp As stdole.IPictureDisp) As Image
                Return GetPictureFromIPicture(pictureDisp)
            End Function
        End Class

    End Class

End Namespace

File: InventorAddInBrowserPanAttempt4/InventorAddInBrowserPanAttempt4/StandardAddInServer.cs
using System;
using System.Runtime.InteropServices;
using Inventor;
using Microsoft.Win32;

namespace InventorAddInBrowserPanAttempt4
{
    /// <summary>
    /// This is the primary AddIn Server class that implements the ApplicationAddInServer interface
    /// that all Inventor AddIns are required to implement. The communication between Inventor and
    /// the AddIn is via the methods on this interface.
    /// </summary>
    [GuidAttribute("153c54e3-f9bc-4759-8f4e-f4f5c2624d78")]
    public class StandardAddInServer : Inventor.ApplicationAddInServer
    {

        // Inventor application object.
        private Inventor.Application m_inventorApplication;
        private ButtonDefinition mAsmButtonDef;
        private ButtonDefinition mPartButtonDef;
        private String strAddInGuid = "e383aeef-321c-42c1-8ed0-3ab07dc33c5d";

        public StandardAddInServer()
        {
        }

        #region ApplicationAddInServer Members

        public void Activate(Inventor.ApplicationAddInSite addInSiteObject, bool firstTime)
        {
            // This method is called by Inventor when it loads the addin.
            // The AddInSiteObject provides access to the Inventor Application object.
            // The FirstTime flag indicates if the addin is loaded for the first time.

            // Initialize AddIn members.
            m_inventorApplication = addInSiteObject.Application;
            ControlDefinitions mCtrlDefs = m_inventorApplication.CommandManager.ControlDefinitions;
            // TODO: Add ApplicationAddInServer.Activate implementation.
            // e.g. event initialization, command creation etc.
            IPictureDisp mIPictureDisp1 = _
             PictureDispConverter.ToIPictureDisp _
                                    (My.Resources.Tools);
            IPictureDisp mIPictureDisp2 = _
             PictureDispConverter.ToIPictureDisp _
                                    (My.Resources.wbTest);
            mAsmButtonDef = mCtrlDefs.AddButtonDefinition("Assembly Cmd", "Autodesk:RibbonVBTest:Button1", CommandTypesEnum.kQueryOnlyCmdType,
            strAddInGuid, "Description", "Tooltip text for Assembly", mIPictureDisp1, mIPictureDisp1,
            ButtonDisplayEnum.kDisplayTextInLearningMode);
            mPartButtonDef = mCtrlDefs.AddButtonDefinition("Part Cmd", "Autodesk:RibbonVBTest:Button2", CommandTypesEnum.kQueryOnlyCmdType,
            strAddInGuid, "Description", "Tooltip text for Part", mIPictureDisp2, mIPictureDisp2,
            ButtonDisplayEnum.kDisplayTextInLearningMode);
            if (firstTime)
            {
                UserInterfaceManager UIManager = m_inventorApplication.UserInterfaceManager;
                if(UIManager.InterfaceStyle == InterfaceStyleEnum.kRibbonInterface){
                    Inventor.Ribbon assemblyRibbon = UIManager.Ribbons.Item("Assembly");
                    Inventor.RibbonTab assemblyTab = assemblyRibbon.RibbonTabs.Item("id_TabAssemble");
                    Inventor.RibbonPanel panel1 = assemblyTab.RibbonPanels.Add("Custom commands",
                        "Autodesk:RibbonVBTest:Panel1",strAddInGuid);
                    panel1.CommandControls.AddButton(mAsmButtonDef, true);
                    Inventor.Ribbon partRibbon = UIManager.Ribbons.Item("Part");
                    Inventor.RibbonTab partTab = assemblyRibbon.RibbonTabs.Item("id_TabModel");
                    Inventor.RibbonPanel panel2 = modelTab.RibbonPanels.Add("Custom commands",
                                                                                     "Autodes:RibbonVBTest:Panel2",
                                                                                       strAddInGuid);
                    panel2.CommandControls.AddButton(mPartButtonDef, true);
                } else
                {
                    CommandBar mAsmCommandBar = UIManager.CommandBars("AMxAssemblyPanelCmdBar");
                    mAsmCommandBar.Controls.AddButton(mAsmButtonDef, 0);
                    CommandBar mPartCommandBar = UIManager.CommandBars("PMxPartFeatureCmdBar");
                        mPartCommandBar.Controls.AddButton(mPartButtonDef, 0);
                }
            }

        }

        public void Deactivate()
        {
            // This method is called by Inventor when the AddIn is unloaded.
            // The AddIn will be unloaded either manually by the user or
            // when the Inventor session is terminated

            // TODO: Add ApplicationAddInServer.Deactivate implementation

            // Release objects.
            m_inventorApplication = null;

            GC.Collect();
            GC.WaitForPendingFinalizers();
        }

        public void ExecuteCommand(int commandID)
        {
            // Note:this method is now obsolete, you should use the 
            // ControlDefinition functionality for implementing commands.
        }

        public object Automation
        {
            // This property is provided to allow the AddIn to expose an API 
            // of its own to other programs. Typically, this  would be done by
            // implementing the AddIn's API interface in a class and returning 
            // that class object through this property.

            get
            {
                // TODO: Add ApplicationAddInServer.Automation getter implementation
                return null;
            }
        }

        #endregion

    }
}
